main.py

```diff
@@ -144,14 +145,96 @@ async def request_termination_of_experiment(rid: int):
     remote.request_termination(rid)
 
 
+def add_tracking_line(stmt_list: List[ast.stmt]) -> List[ast.stmt]:
+    """Modifies the given statements list to leave rtio logs that track each line.
+
+    Adds logs containing the corresponding line number before each line, recursively.
+
+    Args:
+        stmt_list: The list of ast statements.
+
+    Returns:
+        The list of modified ast statements.
+    """
+    modified_stmt_list = []
+    for stmt in stmt_list:
+        tracking_line_code = f"rtio_log('line', {stmt.lineno})"
+        tracking_line_stmt = ast.parse(tracking_line_code).body
+        modified_stmt_list.extend(tracking_line_stmt)
+        for attr in ("body", "handlers", "orelse", "finalbody"):
+            if hasattr(stmt, attr):
+                setattr(stmt, attr, add_tracking_line(getattr(stmt, attr)))
+        modified_stmt_list.append(stmt)
+    return modified_stmt_list
+
+
+def modify_experiment_code(code: str, experiment_cls_name: str) -> str:
+    """Modifies the given code to leave rtio logs and make the vcd file.
+
+    It performs the following modifications:
+        1. Add the rtio logs to track each line.
+        2. Make the vcd file into the corresponding path.
+
+    Args:
+        code: The experiment code.
+        experiment_cls_name: The class name of the experiment.
+
+    Returns:
+        The modified experiment code.
+    """
+    code_mod = ast.parse(code)
+    # import modules
+    import_code = """
+import subprocess
+import posixpath
+    """
+    import_stmt = ast.parse(import_code).body
+    code_mod.body = import_stmt + code_mod.body
+    experiment_cls_stmt = next(
+        stmt for stmt in code_mod.body
+        if isinstance(stmt, ast.ClassDef) and stmt.name == experiment_cls_name
+    )
+    run_func_stmt = next(
+        stmt for stmt in experiment_cls_stmt.body
+        if isinstance(stmt, ast.FunctionDef) and stmt.name == "run"
+    )
+    # make logs to track each line
+    run_func_stmt.body = add_tracking_line(run_func_stmt.body)
+    # call write_vcd()
+    write_vcd_call_stmt = ast.parse("self.write_vcd()").body
+    run_func_stmt.body.append(write_vcd_call_stmt)
+    # define write_vcd()
+    write_vcd_func_code = f"""
+def write_vcd(self):
+    result = subprocess.run("curl http://127.0.0.1:8000/experiment/running/",
+                            capture_output=True, shell=True, text=True)
+    rid = int(result.stdout)
+    device_db_path = posixpath.join(
+        "{configs["master_path"]}",
+        "device_db.py"
+    )
```

Comment:
You can `join()` it in advance and give the result path as a constant string.
Then we don't have to import `posixpath` in the experiment file, do we?

Reply:
Oh, you're right! I will fix it.

Reply:
I realize now that we need to import `posixpath` in order to define `vcd_path`.
It uses the RID so that we cannot define `vcd_path` in advance.

What do you think of it? @kangz12345

Reply:
Even though RID cannot be determined in advance, you can still join the path without knowing it: e.g., `"dir1/dir2/{RID}/..."`.

Reply:
Wow, I see. I will apply your idea!